proposed/request-handler-factory.md

```diff
@@ -0,0 +1,56 @@
+# Request Handler Factory
+
+## Summary
+
+In order to be able to work with other routing and request handling libraries, 
+I feel that that we need a standard factory method that can be used to build 
+the `RequestHandlerInterface` object. 
+
+We have this for `ServerRequestFactoryInterface` and `ResponseFactoryInterface`
+which are very helpful. 
+
+## Interface
+
+I propose the following interface:
+
+```php
+namespace Psr\Http\Message;
+
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+
+interface RequestHandlerFactoryInterface
+{
+    /**
+     * Create a new RequestHandler from the ServerRequest
+     *
+     * @param ServerRequestInterface $serverRequest
+     * @return RequestHandlerInterface
+     */
+    public function createRequestHandler(ServerRequestInterface $serverRequest): RequestHandlerInterface;
+}
+```
+
+Here is an example implementation
+
+```php
+class RequestHandlerFactory implements RequestHandlerFactoryInterface
+{
+    private Router $router;
+
+    public function __construct(Router $router)
+    {
+        $this->router = $router;
+    }
+
+    public function createRequestHandler(ServerRequestInterface $serverRequest): RequestHandlerInterface
+    {
+        $route = $this->router->parseRequest($serverRequest);
+
+        $serverRequestHandler = new Dispatcher();
+        $serverRequesthandler->addMiddlewares($route->getMiddlewares());
+
+        return $serverRequestHandler->addMiddleware(new RoutingMiddleware($route));
+    }
+}
```

Comment:
This is essentially an implementation of a request handler, but called something else. It could also be written as follows (see below), which is how most PSR-15 oriented application frameworks (eg: Slim Framework) are written. The stuff with adding middleware is basically just application bootstrapping, which is generally written in a "factory" class.

```suggestion
class Application implements RequestHandlerInterface
{
    public function __construct(
        private Router $router,
        private ContainerInterface $container,
    ) {
    }

    public function handle(ServerRequestInterface $request): ResponseInterface
    {
        $route = $this->router->parseRequest($serverRequest);
        $handler = $route->getHandler();
        
        return $this->container->get($handler)->handle($request);
    }
}
```

Reply:
I understand but this https://github.com/relayphp/Relay.Relay * is also request handler.

* not used it, but it looks good for the example.

Reply:
Maybe the `getHandler` is confusing for this example, it is more `getCallable` which is going to produce the response e.g. [$controller, 'index'] ,

But the bottom line is for me to be able to use other peoples implementations of PSR-15, the routing needs to be embedded into their implementation, more specifically the handle method.

What I am trying to say is that without a factory, I would have to do something like this I guess, which is creating a request handler which calls another request handler or I have to not use other peoples packages and just create my own to deal with this problem. Both of these solutions seem incorrect.

![image](https://user-images.githubusercontent.com/20553479/140642771-677779ce-bd8f-44c0-9c49-f3ead1bb5c03.png)

Perhaps the factory does not make sense if the middleware building is built when the handle method  is called (as you suggest), but there are many correct implementations of PSR-15 that do not do that, so essentially despite them being PSR-15, I cant just drop them in my projects and use them, I still need to do some factory stuff.

I appreciate you are the editor of the PSR-15 standard, so nobody will no better than you,  and please correct me if I am wrong. My understanding was that RequestHandler was to execute the middleware and return a response, without a standard way to build the middleware stack you cant share the request handler between projects without making changes to code etc. 

Building the middleware stack for PSR-15 requests seems to be done in two ways, either when you call the `handle` method which parses the request or by adding them manually, such as `addMiddleware` method.  

If both are correct then surely we we need a factory?